### PR TITLE
feat: support show server doc description in LspInfo

### PR DESCRIPTION
**Pull request by glepnir**
this is useful  for user to check the server doc in neovim when there is something wrong in server .

images

### CODE REVIEW

**Comment by glepnir** [comment]
@williamboman  how do you feel like this ?

**Comment by williamboman** [comment]
This looks very interesting! I think there are some nil-checks missing, see below for suggestion:

```diff
diff --git a/lua/lspconfig/ui/lspinfo.lua b/lua/lspconfig/ui/lspinfo.lua
index 88894ce..d2b090b 100644
--- a/lua/lspconfig/ui/lspinfo.lua
+++ b/lua/lspconfig/ui/lspinfo.lua
@@ -309,17 +309,24 @@ return function()
 
   local function show_doc()
     local lines = {}
+
+    local function append_lines(config)
+      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
+      if desc then
+        vim.list_extend(lines, vim.split(desc, '\n'))
+        table.insert(lines, '')
+      end
+    end
+
     for _, client in pairs(buf_clients) do
       local config = require('lspconfig.configs')[client.name]
-      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
-      vim.list_extend(lines, vim.split(desc, '\n'))
-      table.insert(lines, '')
+      if config then
+        append_lines(config)
+      end
     end
 
     for _, config in pairs(other_matching_configs) do
-      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
-      vim.list_extend(lines, vim.split(desc, '\n'))
-      table.insert(lines, '')
+      append_lines(config)
     end
 
     local info = windows.percentage_range_window(0.8, 0.7)

```

File: lua/lspconfig/ui/lspinfo.lua [hunk 1]
```diff
@@ -202,7 +202,7 @@ return function()
   end
 
   -- insert the tips at the top of window
-  table.insert(buf_lines, 'Use [q] or [Esc] to quit the window')
+  table.insert(buf_lines, 'Server Doc <TAB> use q or <esc> to quit')
```
Comment (williamboman) [inline lua/lspconfig/ui/lspinfo.lua:205]:
```suggestion
  table.insert(buf_lines, 'Press <Tab> to view server documentation. Press q or <Esc> to close this window.')
```

**Review by williamboman — COMMENTED** [review]
I think the documentation window could need some more work, it wasn't clear to me how to leave it. I personally first tried to press `<Tab>` to go back to the `:LspInfo` landing window, maybe that keybind could be added (on top of `q` and `<Esc>`)?

I also found the server documentation window to be pretty difficult to read. There's a lot going on in it so I think we need to visualize the different sections a bit better, let me experiment with it a bit and I'll come up with some concrete suggestions. I also think some padding in the buffer would help a lot!

**Comment by williamboman** [comment]
After playing around with it for a short while I'd like to suggest the following changes:

```diff
diff --git a/lua/lspconfig/ui/lspinfo.lua b/lua/lspconfig/ui/lspinfo.lua
index 88894ce..617501d 100644
--- a/lua/lspconfig/ui/lspinfo.lua
+++ b/lua/lspconfig/ui/lspinfo.lua
@@ -202,7 +202,7 @@ return function()
   end
 
   -- insert the tips at the top of window
-  table.insert(buf_lines, 'Server Doc <TAB> use q or <esc> to quit')
+  table.insert(buf_lines, 'Press q or <Esc> to close this window. Press <Tab> to view server documentation.')
 
   local header = {
     '',
@@ -309,25 +309,41 @@ return function()
 
   local function show_doc()
     local lines = {}
+
+    local function append_lines(config)
+      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
+      if desc then
+        table.insert(lines, string.format('# %s', config.name))
+        table.insert(lines, '')
+        vim.list_extend(lines, vim.split(desc, '\n'))
+        table.insert(lines, '')
+      end
+    end
+
+    table.insert(lines, 'Press <Tab> to view server info.')
+    table.insert(lines, '')
+
     for _, client in pairs(buf_clients) do
       local config = require('lspconfig.configs')[client.name]
-      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
-      vim.list_extend(lines, vim.split(desc, '\n'))
-      table.insert(lines, '')
+      if config then
+        append_lines(config)
+      end
     end
 
     for _, config in pairs(other_matching_configs) do
-      local desc = vim.tbl_get(config, 'document_config', 'docs', 'description')
-      vim.list_extend(lines, vim.split(desc, '\n'))
-      table.insert(lines, '')
+      append_lines(config)
     end
 
     local info = windows.percentage_range_window(0.8, 0.7)
-    api.nvim_buf_set_lines(info.bufnr, 0, -1, false, lines)
+
+    local fmt_buf_doc_lines = indent_lines(lines, ' ')
+    fmt_buf_doc_lines = vim.lsp.util._trim(fmt_buf_doc_lines, {})
+    api.nvim_buf_set_lines(info.bufnr, 0, -1, false, fmt_buf_doc_lines)
+
+    api.nvim_buf_add_highlight(info.bufnr, 0, 'LspInfoTip', 0, 0, -1)
     vim.bo[info.bufnr].filetype = 'markdown'
-    vim.wo[info.win_id].concealcursor = 'niv'
-    vim.wo[info.win_id].conceallevel = 2
-    vim.bo[info.bufnr].syntax = 'on'
+    vim.wo[info.win_id].breakindent = false
+    vim.wo[info.win_id].breakindentopt = ''
 
     local close_doc_win = function()
       if api.nvim_win_is_valid(info.win_id) then
@@ -335,8 +351,7 @@ return function()
       end
     end
 
-    vim.keymap.set('n', 'q', close_doc_win, { buffer = info.bufnr })
-    vim.keymap.set('n', '<ESC>', close_doc_win, { buffer = info.bufnr })
+    vim.keymap.set('n', '<TAB>', close_doc_win, { buffer = info.bufnr })
   end
 
   vim.keymap.set('n', '<TAB>', show_doc, { buffer = true, nowait = true })
```

I'm still not 100% sure this is the right interface to show a lot of markdown content in. I suggest removing `conceallevel` entirely because it hides a lot of important information like links as well as resetting the default window options (`breakindent` et al.).

**Comment by glepnir** [comment]
about markdown i think most of neovim users use treesitter and I believe they installed markdown parser.   the server doc  we need clean.

File: lua/lspconfig/ui/lspinfo.lua [hunk 2]
```diff
@@ -202,7 +202,7 @@ return function()
   end
 
   -- insert the tips at the top of window
-  table.insert(buf_lines, 'Use [q] or [Esc] to quit the window')
+  table.insert(buf_lines, 'Press q or <Esc> to close this window. Press <Tab> to view server doc.')
```
Comment (justinmk) [inline lua/lspconfig/ui/lspinfo.lua:205]: 👍  good change, follows common conventions.

File: lua/lspconfig/ui/lspinfo.lua [hunk 3]
```diff
@@ -268,7 +268,7 @@ return function()
 
   local function close()
     api.nvim_clear_autocmds { group = augroup, buffer = bufnr }
-    if api.nvim_buf_is_valid(bufnr) then
+    if api.nvim_buf_is_valid(bufnr) and api.nvim_buf_is_loaded(bufnr) then
```
Comment (justinmk) [inline lua/lspconfig/ui/lspinfo.lua:271]: don't need `nvim_buf_is_valid` if check is_loaded

File: lua/lspconfig/ui/lspinfo.lua [hunk 4]
```diff
@@ -306,4 +306,58 @@ return function()
...
+    for _, client in pairs(buf_clients) do
+      local config = require('lspconfig.configs')[client.name]
+      if config then
+        append_lines(config)
+      end
+    end
+
+    for _, config in pairs(other_matching_configs) do
+      if config then
+        if config then
```
Comment (justinmk) [inline lua/lspconfig/ui/lspinfo.lua:334]: duplicate line

File: lua/lspconfig/ui/lspinfo.lua [hunk 5]
```diff
@@ -306,4 +306,58 @@ return function()
...
+      end
+    end
+
+    table.insert(lines, 'Press <Tab> to close server info.')
+    table.insert(lines, '')
+
+    for _, client in pairs(buf_clients) do
+      local config = require('lspconfig.configs')[client.name]
+      if config then
+        append_lines(config)
```
Comment (justinmk) [inline lua/lspconfig/ui/lspinfo.lua:328]: append_lines could accept `nil`, that avoids boilerplate in the callers

**Comment by justinmk** [comment]
screenshot? 

could this potentially live in `:checkhealth vim.lsp` instead of adding more stuff to LspInfo ?

**Comment by glepnir** [comment]
updated screenshot

**Pull request merged**